flips/component-interface.md

```diff
@@ -0,0 +1,267 @@
+      close(ready)
+    }()
+
+    return ready
+  }
+
+  func (c *ComponentManager) Done() <-chan struct{} {
+    done := make(chan struct{})
+
+    go func() {
+      <-lifecycle.AllDone(c.subComponents...)
+      c.childRoutinesFinished.Wait()
+      close(done)
+    }()
+
+    return done
+  }
+  ```
+
+  Components that want to implement `Component` can use this `ComponentManager` to simplify implementation:
+
+  ```golang
+  type FooComponent struct {
+    *ComponentManager
+  }
+
+  func NewFooComponent(foo fooType, bar barType) *FooComponent {
+    f := &FooComponent{}
+
+    cmb := NewComponentManagerBuilder().
+      AddChildRoutine(f.childRoutine).
+      AddChildRoutine(f.childRoutineWithFooParameter(foo)).
+      AddSubComponent(NewBarComponent(bar))
+
+    f.ComponentManager = cmb.Build()
+
+    return f
+  }
+
+  func (f *FooComponent) childRoutine(ctx context.Context) {
+    for {
+      select {
+      case <-ctx.Done():
+        return
+      default:
+        // do work...
+      }
+    }
+  }
+
+  func (f *FooComponent) childRoutineWithFooParameter(foo fooType) func(context.Context) {
+    return func(ctx context.Context) {
+      for {
+        select {
+        case <-ctx.Done():
+          return
+        default:
+          // do work with foo...
+        }
+      }
+    }
+  }
+  ```
+
+* `ErrorAware` can be implemented with the help of an `ErrorManager` struct:
+
+  ```golang
+  // ErrorManager implements the ErrorAware interface, and provides a way for components
+  // to signal an irrecoverable error.
+  type ErrorManager struct {
+    errors chan error
+  }
+
+  func NewErrorManager() *ErrorManager {
+    return &ErrorManager{make(chan error)}
+  }
+
+  func (e *ErrorManager) Errors() <-chan error {
+    return e.errors
+  }
+
+  func (e *ErrorManager) ThrowError(err error) {
+    e.errors <- err
+    runtime.Goexit()
+  }
+  ```
+
+  Error handling for components can be implemented like so:
+
+  ```golang
+  type ErrorHandler func(ctx context.Context, errors <-chan error, restart func())
+
+  type ComponentFactory func() (Component, error)
+
+  func RunComponent(ctx context.Context, componentFactory ComponentFactory, errorHandler ErrorHandler) error {
+    restartChan := make(chan struct{})
+
+    start := func() (context.CancelFunc, <-chan struct{}, error) {
+      component, err := componentFactory()
+      if err != nil {
+        // failed to create component
+        return nil, nil, err
+      }
+
+      // context used to restart the component
+      runCtx, cancel := context.WithCancel(ctx)
+      if err := component.Start(runCtx); err != nil {
+        // failed to start component
+        cancel()
+        return nil, nil, err
+      }
+
+      select {
+      case <-ctx.Done():
+        return nil, nil, ctx.Err()
+      case <-component.Ready():
+      }
+
+      go errorHandler(runCtx, component.Errors(), func() {
+        restartChan <- struct{}{}
+        runtime.Goexit()
+      })
```

Comment:
As you said, it doesn't actually matter much. The reason I started it after is because we want to make sure the component actually starts up properly, otherwise we don't even need to consider this part.